subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/FilteredModuleComponentRepository.java

```diff
@@ -0,0 +1,207 @@
+    private final String consumerName;
+    private final ImmutableAttributes consumerAttributes;
+
+    public static ModuleComponentRepository of(ModuleComponentRepository delegate, Action<? super ArtifactResolutionDetails> action, String consumerName, AttributeContainer attributes) {
+        if (action == null) {
+            return delegate;
+        }
+        return new FilteredModuleComponentRepository(delegate, action, consumerName, attributes);
+    }
+
+    private FilteredModuleComponentRepository(ModuleComponentRepository delegate, Action<? super ArtifactResolutionDetails> filterAction, String consumerName, AttributeContainer attributes) {
+        this.delegate = delegate;
+        this.filterAction = filterAction;
+        this.consumerName = consumerName;
+        this.consumerAttributes = ((AttributeContainerInternal)attributes).asImmutable();
+    }
+
+
+    @Override
+    public String getId() {
+        return delegate.getId();
+    }
+
+    @Override
+    public String getName() {
+        return delegate.getName();
+    }
+
+    @Override
+    public ModuleComponentRepositoryAccess getLocalAccess() {
+        return new FilteringAccess(delegate.getLocalAccess());
+    }
+
+    @Override
+    public ModuleComponentRepositoryAccess getRemoteAccess() {
+        return new FilteringAccess(delegate.getRemoteAccess());
+    }
+
+    @Override
+    public Map<ComponentArtifactIdentifier, ResolvableArtifact> getArtifactCache() {
+        return delegate.getArtifactCache();
+    }
+
+    @Nullable
+    @Override
+    public InstantiatingAction<ComponentMetadataSupplierDetails> getComponentMetadataSupplier() {
+        return delegate.getComponentMetadataSupplier();
+    }
+
+    private class FilteringAccess implements ModuleComponentRepositoryAccess {
+        private final ModuleComponentRepositoryAccess delegate;
+
+        private FilteringAccess(ModuleComponentRepositoryAccess delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public void listModuleVersions(ModuleDependencyMetadata dependency, BuildableModuleVersionListingResolveResult result) {
+            ModuleIdentifier identifier = dependency.getSelector().getModuleIdentifier();
+            whenModulePresent(identifier, null,
+                    () -> delegate.listModuleVersions(dependency, result),
+                    () -> result.listed(Collections.emptyList()));
+        }
+
+        @Override
+        public void resolveComponentMetaData(ModuleComponentIdentifier moduleComponentIdentifier, ComponentOverrideMetadata requestMetaData, BuildableModuleComponentMetaDataResolveResult result) {
+            whenModulePresent(moduleComponentIdentifier.getModuleIdentifier(), moduleComponentIdentifier,
+                    () -> delegate.resolveComponentMetaData(moduleComponentIdentifier, requestMetaData, result),
+                    () -> result.missing());
+        }
```

Comment:
I don't think this pattern should work: if we do a version listing for `org:foo:1.+` and choose the version `1.2`, then we should _expect_ the version 1.2 to be in the same repository. So if someone filters out this version, then the resolution should fail rather than move onto the next repository. 

Even though our current implementation doesn't enforce this separation, I don't think we should expose these as separate actions in the API. This will lead to confusion when someone uses a content filter that excludes a particular version (by component id) but that exact version is resolved from the repository via a dynamic version selector.

This is another vote for having a single filter based on `ModuleComponentSelector` rather than trying to filter on the actual component identifier (which has already been "resolved" in this repository).